### PR TITLE
DMP-1891: Endpoint to delete annotations

### PR DESCRIPTION
**Pull request by tonymort**
Endpoint to delete annotations

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/resources/openapi/annotations.yaml [hunk 1]
```diff
@@ -47,6 +47,25 @@ paths:
...
+      parameters:
+        - in: path
+          name: annotation_id
+          schema:
+            type: integer
+          required: true
+      responses:
+        '204':
+          description: Annotation has been deleted
+        '403':
```
Comment (cmagowan-hmcts) [inline src/main/resources/openapi/annotations.yaml:65]: should we be including references to the problem.yaml here?

File: src/main/java/uk/gov/hmcts/darts/annotation/controller/AnnotationController.java [hunk 2]
```diff
@@ -27,8 +29,17 @@ public class AnnotationController implements AnnotationsApi {
...
     @Override
-    public ResponseEntity<PostAnnotationResponse> annotationsPost(MultipartFile file, Annotation annotation) {
+    public ResponseEntity<PostAnnotationResponse> postAnnotation(MultipartFile file, Annotation annotation) {
         var annotationId = annotationService.process(file, annotation);
         return ResponseEntity.ok(new PostAnnotationResponse(annotationId));
     }
+
+    @Authorisation(
+        contextId = ANY_ENTITY_ID,
+        globalAccessSecurityRoles = {JUDGE, ADMIN})
```
Comment (cmagowan-hmcts) [inline src/main/java/uk/gov/hmcts/darts/annotation/controller/AnnotationController.java:39]: probably want to have `securityRoles = {JUDGE},` here also, as judges can be on probation where they dont have global access

Comment (tonymort) [reply src/main/java/uk/gov/hmcts/darts/annotation/controller/AnnotationController.java:39]: I did check this with Dave, he said not.  Possible I misunderstood so will get his input on this

File: src/main/resources/openapi/annotations.yaml (hunk 1 above)
Comment (tonymort) [reply src/main/resources/openapi/annotations.yaml:65]: possibly, will take a look

File: src/main/java/uk/gov/hmcts/darts/annotation/controller/AnnotationController.java [hunk 3]
```diff
@@ -27,8 +29,17 @@ public class AnnotationController implements AnnotationsApi {
...
         globalAccessSecurityRoles = {JUDGE})
     @Override
-    public ResponseEntity<PostAnnotationResponse> annotationsPost(MultipartFile file, Annotation annotation) {
+    public ResponseEntity<PostAnnotationResponse> postAnnotation(MultipartFile file, Annotation annotation) {
         var annotationId = annotationService.process(file, annotation);
         return ResponseEntity.ok(new PostAnnotationResponse(annotationId));
     }
+
+    @Authorisation(
+        contextId = ANY_ENTITY_ID,
```
[the remark this answers is not in the document]
Comment (tonymort) [reply src/main/java/uk/gov/hmcts/darts/annotation/controller/AnnotationController.java:38]: Spoke to Hemanta about this, he said not.  Again will double check this.

File: src/main/java/uk/gov/hmcts/darts/annotation/component/UserAuthorisedToDeleteAnnotationValidator.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+public class UserAuthorisedToDeleteAnnotationValidator implements Validator<Integer> {
+
+    private final AnnotationRepository annotationRepository;
+    private final UserIdentity userIdentity;
+
+    @Override
+    public void validate(Integer id) {
+        var annotation = annotationRepository.findById(id).orElseThrow(() -> new DartsApiException(ANNOTATION_NOT_FOUND));
+        var currentUser = userIdentity.getUserAccount();
+        var currentUserRoles = securityRolesOf(currentUser);
```
Comment (cmagowan-hmcts) [inline src/main/java/uk/gov/hmcts/darts/annotation/component/UserAuthorisedToDeleteAnnotationValidator.java:28]: not sure if you want to use AuthorisationApiImpl.userHasOneOfRoles instead?

File: src/main/java/uk/gov/hmcts/darts/authorisation/component/impl/AuthorisationImpl.java [hunk 5]
```diff
@@ -166,6 +168,20 @@ public void authoriseTransformedMediaAgainstUser(Integer transformedMediaId) {
         checkMediaRequestIsRequestedByUser(transformedMediaRepository.getReferenceById(transformedMediaId).getMediaRequest());
     }
 
+    @Override
+    public void authoriseByAnnotationId(Integer annotationId, Set<SecurityRoleEnum> securityRoles) {
+        try {
+            var courthouses = annotationRepository.getReferenceById(annotationId).getHearingList().stream()
```
Comment (cmagowan-hmcts) [inline src/main/java/uk/gov/hmcts/darts/authorisation/component/impl/AuthorisationImpl.java:174]: should this not be findById?

File: src/main/java/uk/gov/hmcts/darts/common/repository/AnnotationRepository.java [hunk 6]
```diff
@@ -53,4 +55,7 @@ public interface AnnotationRepository extends JpaRepository<AnnotationEntity, In
         """)
     List<AnnotationEntity> findByListOfHearingIdsAndUser(List<Integer> hearingIds, UserAccountEntity userAccount);
 
+    @Modifying
+    @Query("update AnnotationEntity ann set ann.deleted = true where ann.id = :id")
```
Comment (cmagowan-hmcts) [inline src/main/java/uk/gov/hmcts/darts/common/repository/AnnotationRepository.java:59]: probably want to update the last_modified_by also

File: src/main/resources/openapi/annotations.yaml [hunk 7]
```diff
@@ -50,6 +50,33 @@ paths:
             application/json+problem:
               schema:
                 $ref: './problem.yaml'
+  /annotations/{annotation_id}:
```
Comment (cmagowan-hmcts) [inline src/main/resources/openapi/annotations.yaml:53]: Can you add new endpoints to the postman collection?

File: src/main/java/uk/gov/hmcts/darts/annotation/component/UserAuthorisedToDeleteAnnotationValidator.java (hunk 4 above)
Comment (tonymort) [reply src/main/java/uk/gov/hmcts/darts/annotation/component/UserAuthorisedToDeleteAnnotationValidator.java]: done

File: src/main/java/uk/gov/hmcts/darts/authorisation/component/impl/AuthorisationImpl.java (hunk 5 above)
Comment (tonymort) [reply src/main/java/uk/gov/hmcts/darts/authorisation/component/impl/AuthorisationImpl.java]: done, will look into most efficient way as a background task

File: src/main/java/uk/gov/hmcts/darts/common/repository/AnnotationRepository.java (hunk 6 above)
Comment (tonymort) [reply src/main/java/uk/gov/hmcts/darts/common/repository/AnnotationRepository.java]: done

File: src/main/resources/openapi/annotations.yaml (hunk 7 above)
Comment (tonymort) [reply src/main/resources/openapi/annotations.yaml:53]: done

**Review by cmagowan-hmcts — APPROVED** [review]

**Pull request merged**